robotbase/src/main/java/com/team2357/frc2022/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,147 @@
+package com.team2357.frc2022.subsystems;
+
+import com.revrobotics.CANSparkMax;
+import com.revrobotics.SparkMaxPIDController;
+import com.revrobotics.CANSparkMax.IdleMode;
+import com.team2357.lib.subsystems.ClosedLoopSubsystem;
+import com.team2357.frc2022.util.Utility;
+
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+
+public class TurretSubsystem extends ClosedLoopSubsystem {
+    CANSparkMax m_turretMotor;
+
+    private SparkMaxPIDController m_pidController;
+
+    private Configuration m_config;
+
+    private double m_targetMotorRotations = Double.NaN;
+
+    public static class Configuration {
+        public IdleMode turretMotorIdleMode = IdleMode.kBrake;
+        public int m_turretMotorStallLimitAmps = 0;
+        public int m_turretMotorFreeLimitAmps = 0;
+
+        public double m_turretMotorP = 0;
+        public double m_turretMotorI = 0;
+        public double m_turretMotorD = 0;
+        public double m_turretMotorIZone = 0;
+        public double m_turretMotorFF = 0;
+        public double m_turretMotorMaxOutput = 0;
+        public double m_turretMotorMinOutput = 0;
+        public double m_turretMotorMaxRPM = 0;
+        public double m_turretMotorMaxVel = 0;
+        public double m_turretMotorMinVel = 0;
+        public double m_turretMotorMaxAcc = 0;
+        public double m_turretMotorAllowedError = 0;
+
+        public double m_turretRotationsCounterClockwiseSoftLimit = 0.0;
+        public double m_turretRotationsClockwiseSoftLimit = 0.0;
+        public double m_turretGearRatio = 0.0;
+    }
+
+    public TurretSubsystem(CANSparkMax turretMotor) {
+        m_turretMotor = turretMotor;
+        m_turretMotor.getEncoder().setPosition(0);
+    }
+
+    public void configure(Configuration config) {
+        m_config = config;
+
+        m_turretMotor.setIdleMode(m_config.turretMotorIdleMode);
+        m_turretMotor.setSmartCurrentLimit(m_config.m_turretMotorStallLimitAmps, m_config.m_turretMotorFreeLimitAmps);
+
+        m_pidController = m_turretMotor.getPIDController();
+
+        // set PID coefficients
+        m_pidController.setP(m_config.m_turretMotorP);
+        m_pidController.setI(m_config.m_turretMotorI);
+        m_pidController.setD(m_config.m_turretMotorD);
+        m_pidController.setIZone(m_config.m_turretMotorIZone);
+        m_pidController.setFF(m_config.m_turretMotorFF);
+        m_pidController.setOutputRange(m_config.m_turretMotorMinOutput, m_config.m_turretMotorMinOutput);
+
+        // Configure smart motion
+        int smartMotionSlot = 0;
+        m_pidController.setSmartMotionMaxVelocity(m_config.m_turretMotorMaxVel, smartMotionSlot);
```

Comment:
FYI, I know the name is misleading, but this is the function that sets the cruise velocity of the trapezoidal curve.